subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c

```diff
@@ -156,158 +163,201 @@ do_firmware_transfer_reply_cb(const struct zoap_packet *response,
 
-static enum zoap_block_size default_block_size(void)
-{
-	switch (CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_COAP_BLOCK_SIZE) {
-	case 16:
-		return ZOAP_BLOCK_16;
-	case 32:
-		return ZOAP_BLOCK_32;
-	case 64:
-		return ZOAP_BLOCK_64;
-	case 128:
-		return ZOAP_BLOCK_128;
-	case 256:
-		return ZOAP_BLOCK_256;
-	case 512:
-		return ZOAP_BLOCK_512;
-	case 1024:
-		return ZOAP_BLOCK_1024;
-	}
-
-	return ZOAP_BLOCK_256;
-}
-
 static void firmware_transfer(struct k_work *work)
 {
-#if defined(CONFIG_NET_IPV6)
-	static struct sockaddr_in6 any_addr6 = { .sin6_addr = IN6ADDR_ANY_INIT,
-						.sin6_family = AF_INET6 };
-#endif
-#if defined(CONFIG_NET_IPV4)
-	static struct sockaddr_in any_addr4 = { .sin_addr = INADDR_ANY_INIT,
-					       .sin_family = AF_INET };
-#endif
-	struct net_if *iface;
-	int ret, port, family;
+	int ret, family;
+	u16_t off;
+	u16_t len;
+	char tmp;
 
 	/* Server Peer IP information */
-	/* TODO: use parser on firmware_uri to determine IP version + port */
-	/* TODO: hard code IPv4 + port for now */
-	port = 5685;
 	family = AF_INET;
 
-#if defined(CONFIG_NET_IPV6)
-	if (family == AF_INET6) {
-		firmware_addr.sa_family = family;
-		/* HACK: use firmware_uri directly as IP address */
-		net_addr_pton(firmware_addr.sa_family, firmware_uri,
-			      &net_sin6(&firmware_addr)->sin6_addr);
-		net_sin6(&firmware_addr)->sin6_port = htons(5685);
+	http_parser_url_init(&parsed_uri);
+	ret = http_parser_parse_url(firmware_uri,
+				    strlen(firmware_uri),
+				    0,
+				    &parsed_uri);
+	if (ret != 0) {
+		SYS_LOG_ERR("Invalid firmware URI: %s", firmware_uri);
+		lwm2m_firmware_set_update_result(RESULT_INVALID_URI);
+		return;
 	}
-#endif
 
-#if defined(CONFIG_NET_IPV4)
-	if (family == AF_INET) {
-		firmware_addr.sa_family = family;
-		net_addr_pton(firmware_addr.sa_family, firmware_uri,
-			      &net_sin(&firmware_addr)->sin_addr);
-		net_sin(&firmware_addr)->sin_port = htons(5685);
+	/* Check schema and only support coap for now */
+	if (!(parsed_uri.field_set & (1 << UF_SCHEMA))) {
+		SYS_LOG_ERR("No schema in package uri");
+		lwm2m_firmware_set_update_result(RESULT_INVALID_URI);
+		return;
 	}
-#endif
 
-	ret = net_context_get(firmware_addr.sa_family, SOCK_DGRAM, IPPROTO_UDP,
-			      &firmware_net_ctx);
-	if (ret) {
-		NET_ERR("Could not get an UDP context (err:%d)", ret);
+	/* TODO: enable coaps when DTLS is ready */
+	off = parsed_uri.field_data[UF_SCHEMA].off;
+	len = parsed_uri.field_data[UF_SCHEMA].len;
+	if (len != 4 || memcmp(firmware_uri + off, "coap", 4)) {
+		SYS_LOG_ERR("Unsupported schema");
+		lwm2m_firmware_set_update_result(RESULT_UNSUP_PROTO);
 		return;
 	}
 
-	iface = net_if_get_default();
-	if (!iface) {
-		NET_ERR("Could not find default interface");
-		goto cleanup;
+	if (!(parsed_uri.field_set & (1 << UF_PORT))) {
+		/* Set to default port of CoAP */
+		parsed_uri.port = 5683;
+	}
+
+	off = parsed_uri.field_data[UF_HOST].off;
+	len = parsed_uri.field_data[UF_HOST].len;
+
+	/* IPv6 is wrapped by brackets */
```

Comment:
Although these are changes made by me for firmware pull.

line 274~301 & 307 ~ 310 and be removed.
We can pass the host string to net_app_init_udp_client() directly.
Check the patch: https://hastebin.com/qosizocisi.diff